crypto/params.c

```diff
@@ -0,0 +1,365 @@
+        return 1
+
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, sizeof(double));
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+        break;
+    }
+    return 0;    
+#undef CASE
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, char *buf,
+                                            size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string(const char *key, void *buf,
+                                             size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_OCTET_STRING, buf, bsize,
+                                rsize);
+}
+
+static int get_string_internal(const OSSL_PARAM *p, void **val, size_t max_len,
+                               size_t *used_len, unsigned int type)
+{
+    if (val == NULL || p == NULL || p->data_type != type)
+        return 0;
+
+    if (used_len != NULL)
+        *used_len = p->buffer_size;
+
+    if (*val == NULL) {
+        char *const q = OPENSSL_malloc(p->buffer_size);
+
+        if (q == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, p->buffer_size);
+        return 1;
+    }
+    if (max_len < p->buffer_size)
+        return 0;
+    memcpy(*val, p->buffer, p->buffer_size);
+    return 1;
+}
+
+int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, char **val, size_t max_len)
+{
+    return get_string_internal(p, (void **)val, max_len, NULL,
+                               OSSL_PARAM_UTF8_STRING);
+}
+
+int OSSL_PARAM_get_octet_string(const OSSL_PARAM *p, void **val, size_t max_len,
+                                size_t *used_len)
+{
+    return get_string_internal(p, val, max_len, used_len,
+                               OSSL_PARAM_OCTET_STRING);
+}
+
+static int set_string_internal(OSSL_PARAM *p, const void *val, size_t len,
+                               unsigned int type)
+{
+    SET_RETURN_SIZE(p, len);
+    if (p->data_type == type && p->buffer_size >= len) {
+        memcpy(p->buffer, val, len);
+        return 1;
+    }
+    return 0;
+}
+
+int OSSL_PARAM_set_utf8_string(OSSL_PARAM *p, const char *val)
+{
+    if (p == NULL)
+        return 0;
+    SET_RETURN_SIZE(p, 0);
+    if (val == NULL)
+            return 0;
+    return set_string_internal(p, val, strlen(val) + 1, OSSL_PARAM_UTF8_STRING);
+
+}
+int OSSL_PARAM_set_octet_string(OSSL_PARAM *p, const void *val, size_t len)
+{
+    if (p == NULL)
+        return 0;
+    SET_RETURN_SIZE(p, 0);
+    if (val == NULL)
+            return 0;
```

Comment:
You have these lines duplicated unnessessarily:

``` C
    SET_RETURN_SIZE(p, 0);
    if (val == NULL)
            return 0;
```

Why not have them inside `set_string_internal`?

Reply:
`strlen(val)` is the reason.  The pointer has to be checked for NULL before the length of the utf8 string is calculated or some compilers will remove the NULL check entirely.

Reply:
Ah!  I totally missed that.  Ok then